kani-compiler/src/kani_middle/transform/check_uninit/delayed_ub/instrumentation_visitor.rs

```diff
@@ -63,43 +57,52 @@ impl<'a, 'tcx> InstrumentationVisitor<'a, 'tcx> {
         tcx: TyCtxt<'tcx>,
     ) -> Self {
         Self {
-            skip_next: false,
-            current: SourceInstruction::Statement { idx: 0, bb: 0 },
-            target: None,
+            targets: vec![],
+            current_instruction: SourceInstruction::Statement { idx: 0, bb: 0 },
+            current_target: InitRelevantInstruction {
+                source: SourceInstruction::Statement { idx: 0, bb: 0 },
+                before_instruction: vec![],
+                after_instruction: vec![],
+            },
             points_to,
             analysis_targets,
             current_instance,
             tcx,
         }
     }
+
     fn push_target(&mut self, source_op: MemoryInitOp) {
-        let target = self.target.get_or_insert_with(|| InitRelevantInstruction {
-            source: self.current,
-            after_instruction: vec![],
-            before_instruction: vec![],
-        });
-        target.push_operation(source_op);
+        // If we switched to the next instruction, push the old one onto the list of targets.
+        if self.current_target.source != self.current_instruction {
+            self.targets.push(self.current_target.clone());
+            self.current_target = InitRelevantInstruction {
+                source: self.current_instruction,
+                after_instruction: vec![],
+                before_instruction: vec![],
+            };
+        }
+        self.current_target.push_operation(source_op);
     }
 }
 
 impl<'a, 'tcx> MirVisitor for InstrumentationVisitor<'a, 'tcx> {
     fn visit_statement(&mut self, stmt: &Statement, location: Location) {
-        if self.skip_next {
-            self.skip_next = false;
-        } else if self.target.is_none() {
-            // Check all inner places.
-            self.super_statement(stmt, location);
-        }
+        self.super_statement(stmt, location);
         // Switch to the next statement.
-        let SourceInstruction::Statement { idx, bb } = self.current else { unreachable!() };
-        self.current = SourceInstruction::Statement { idx: idx + 1, bb };
+        if let SourceInstruction::Statement { idx, bb } = self.current_instruction {
+            self.current_instruction = SourceInstruction::Statement { idx: idx + 1, bb }
```

Comment:
don't you need to set the current instruction before visiting the instruction itself? I.e.: before `super_statement` call?

Reply:
It is initially set before `visit_basic_block` is created and then incremented on every iteration:
https://github.com/model-checking/kani/pull/3438/files/58b689a7cd817e9eeb9402f22abb9016911bdc8c#diff-471a7a8b457c94203de190d77f830f88b61e027680d347e9768c24f926a19270R42-R44

I could potentially change the `current_instruction` to be of the `Option<SourceInstruction>` type and initialize it if it is set to None, but it becomes a bit more complicated when we need to account for multiple basic blocks visited.